cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -186,6 +186,12 @@ public final ProgramController run(final Program program, final ProgramOptions o
         public TwillController launch(TwillApplication twillApplication, Iterable<String> extraClassPaths,
                                       Iterable<? extends Class<?>> extraDependencies) {
           TwillPreparer twillPreparer = twillRunner.prepare(twillApplication);
+          // TODO: CDAP-5506. It's a bit hacky to set a Spark environment here. However, we always launch
+          // Spark using YARN and it is needed for both Workflow and Spark runner. We need to set it
+          // because inside Spark code, it will set and unset the SPARK_YARN_MODE system properties, causing
+          // fork in distributed mode not work. By setting it in the environment, which Spark always use it
```

Comment:
nit: Setting it in the environment, which Spark uses for defaults, so it can't be unset by Spark